packages/db-commons/index.cjs

```diff
@@ -86,21 +86,163 @@ const inferColumnTypes = function (rows) {
 	return columnTypes;
 };
 
+/**
+ *
+ * @param {Record<string, unknown>[]} rows
+ * @param {ColumnDefinition[]} columnTypes
+ */
+function applyColumnTypes(rows, columnTypes) {
+	const columns = Object.fromEntries(columnTypes.map((x) => [x.name, x.evidenceType]));
+	for (const row of rows) {
+		for (const column_name in row) {
+			if (columns[column_name] === EvidenceType.BOOLEAN && typeof row[column_name] !== 'boolean') {
+				row[column_name] = Boolean(row[column_name]);
+			} else if (
+				columns[column_name] === EvidenceType.STRING &&
+				typeof row[column_name] !== 'string'
+			) {
+				row[column_name] = JSON.stringify(row[column_name]);
+			} else if (
+				columns[column_name] === EvidenceType.NUMBER &&
+				typeof row[column_name] !== 'number'
+			) {
+				row[column_name] = Number(row[column_name]);
+			} else if (columns[column_name] === EvidenceType.DATE && !(row instanceof Date)) {
+				row[column_name] = new Date(row[column_name]);
+			}
+		}
+	}
+	return rows;
+}
+
+/**
+ * Checks if all the columns of type `evidenceType` in a set of rows
+ * conform to the proper `typeof` or `instanceof`
+ * @param {Record<string, unknown>[]} rows
+ * @param {ColumnDefinition[]} columns
+ * @param {EvidenceType} evidenceType
+ * @returns {boolean}
+ */
+function conformsTo(rows, columns, evidenceType) {
+	if (columns.every(({ evidenceType: columnType }) => columnType !== evidenceType)) return true;
+	if (rows.length === 0) return true;
+
+	const firstRow = rows[0];
+	const columnsUnderInvestigation = columns.filter(
+		({ evidenceType: columnType }) => columnType === evidenceType
+	);
+
+	switch (evidenceType) {
+		case EvidenceType.STRING:
+			return columnsUnderInvestigation.every(({ name }) => typeof firstRow[name] === 'string');
+		case EvidenceType.NUMBER:
+			return columnsUnderInvestigation.every(({ name }) => typeof firstRow[name] === 'number');
+		case EvidenceType.BOOLEAN:
+			return columnsUnderInvestigation.every(({ name }) => typeof firstRow[name] === 'boolean');
+		case EvidenceType.DATE:
+			return columnsUnderInvestigation.every(({ name }) => firstRow[name] instanceof Date);
+	}
+}
+
+/**
+ *
+ * @param {Record<string, unknown>[]} results
+ * @param {ColumnDefinition[]} columns
+ * @returns {Record<string, unknown>[]}
+ */
+function convertStringColumns(results, columns) {
```

Comment:
Is this something that could be handled when loading into parquet files?

Reply:
It could make sense to run it before parquet conversion yeah, arrow just calls `String` on non-strings by default so this would get a bit better compat for objects

Reply:
I don't think we need to worry about support for objects /  dicts yet; we don't handle those very well in other places